### Module 14 - Case study 1

Business challenge/requirement:

FuPont is a leading chemical company across the globe. The Company is on a CSR (Corporate Social Responsibility) mission. It wants to identify biodegradable products based on a study of the relationships between chemical structure and biodegradation of molecules.
You as ML expert have to create a ML model to classify the chemical structure as 'Ready BioDegradable' – RB vs 'Not Ready Biodegradable' – NRB

Key issues:

Data has lots of attributes and classification could be tricky

Data volume:

- Approx 1055 records – file bio-degradabale-data.csv

Business benefits:

Research can lead FuPont to create truly unique Biodegradable packaging material. This could lead to massive profits in future

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [16]:
df = pd.read_csv(r'D:\E\Courses\Edureka\Assignments\Dataset\14. Model Selection and Boosting\bio-degradabale-data.csv', header=None, delimiter=';')

In [17]:
print(df.shape)
df.head()

(1055, 42)


0       1   2   3   4   5   6     7   8   9   ...  32  33  34     35  \
0  3.919  2.6909   0   0   0   0   0  31.4   2   0  ...   0   0   0  2.949   
1  4.170  2.1144   0   0   0   0   0  30.8   1   1  ...   0   0   0  3.315   
2  3.932  3.2512   0   0   0   0   0  26.7   2   4  ...   0   0   1  3.076   
3  3.000  2.7098   0   0   0   0   0  20.0   0   2  ...   0   0   1  3.046   
4  4.236  3.3944   0   0   0   0   0  29.4   2   4  ...   0   0   0  3.351   

      36  37     38  39  40  41  
0  1.591   0  7.253   0   0  RB  
1  1.967   0  7.257   0   0  RB  
2  2.417   0  7.601   0   0  RB  
3  5.000   0  6.690   0   0  RB  
4  2.405   0  8.003   0   0  RB  

[5 rows x 42 columns]

In [21]:
x = df.iloc[:,:41].values
y = df.iloc[:,41].values

In [22]:
x, y

(array([[ 3.919 ,  2.6909,  0.    , ...,  7.253 ,  0.    ,  0.    ],
        [ 4.17  ,  2.1144,  0.    , ...,  7.257 ,  0.    ,  0.    ],
        [ 3.932 ,  3.2512,  0.    , ...,  7.601 ,  0.    ,  0.    ],
        ...,
        [ 4.869 ,  1.767 ,  0.    , ...,  9.537 ,  1.    ,  0.    ],
        [ 5.158 ,  1.6914,  2.    , ..., 11.055 ,  0.    ,  1.    ],
        [ 5.076 ,  2.6588,  2.    , ...,  9.13  ,  0.    ,  2.    ]]),
 array(['RB', 'RB', 'RB', ..., 'NRB', 'NRB', 'NRB'], dtype=object))

In [23]:
from sklearn.preprocessing import LabelEncoder

In [24]:
le = LabelEncoder()
y = le.fit_transform(y)

In [25]:
y

array([1, 1, 1, ..., 0, 0, 0])

In [26]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [27]:
len(x_train), len(x_test), len(y_train), len(y_test)

(844, 211, 844, 211)

In [28]:
from xgboost import XGBClassifier

In [29]:
xgb_classifier = XGBClassifier()
xgb_classifier.fit(x_train, y_train)

XGBClassifier(base_score=0.5, booster=None, colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints=None,
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=0, num_parallel_tree=1,
              objective='binary:logistic', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method=None,
              validate_parameters=False, verbosity=None)

In [31]:
y_pred = xgb_classifier.predict(x_test)

In [32]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [33]:
confusion_matrix(y_test, y_pred)

array([[133,  19],
       [  6,  53]], dtype=int64)

In [34]:
accuracy_score(y_test, y_pred)

0.8815165876777251

In [35]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.88      0.91       152
           1       0.74      0.90      0.81        59

    accuracy                           0.88       211
   macro avg       0.85      0.89      0.86       211
weighted avg       0.90      0.88      0.88       211



88.2% accuracy is with the xgb_classifier model, which is a very good one